doc: Add Redeemer docs

kyriediculous opened the pull request:
**What does this pull request do? Explain your changes. (required)**
This PR adds documentation for the `Redeemer` service and `ticketQueue` table. 

**Does this pull request close any open issues?**
Fixes #1532 


**Checklist:**
- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

Commit:
doc: add ticketQueue table docs

Commit:
doc: Redeemer

yondonfu commented on doc/database.md:
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
+createdAt | DATETIME DEFAULT CURRENT_TIMESTAMP | Time this row was inserted 
+sender | STRING | Address of the broadcaster that sent the winning ticket.
+recipient | STRING | Address of the orchestrator that payments will be credited to.
+faceValue | BLOB | Face value of the ticket, in wei Unique nonce sent by the broadcaster.
```
"Unique nonce sent by the broadcaster" can be removed. I see that this was also in the description for the `winningTickets` table...not sure how that got there 🤷‍♂️

kyriediculous replied:
fixed in e5624db9

yondonfu commented on doc/database.md:
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
+winProb | BLOB | The ticket's winning probability in the range of 0 through 2^256-1.
+senderNonce | INTEGER | Nonce incorporated by the broadcaster with each ticket.
+recipientRand | BLOB | Value used by the orchestrator when constructing the initial ticket parameters.
+recipientRandHash | STRING | Hash of the recipient rand, keccak256(recipientRand).,
```
Let's remove the trailing commas for some of these descriptions.

kyriediculous replied:
fixed in e5624db9

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+2. Anytime a winning ticket is submitted to the `Redeemer Server`, the `Redeemer Server` will monitor it's own `LocalSenderMonitor` for `maxFloat` change notifications for the sender of the ticket using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` (if not already watching for updates from that particular sender). 
+
+3. Whenever the subscription from `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` receives a notificiation, the `Redeemer Server` will request the latest `maxFloat` for the `sender` using `LocalSenderMonitor.MaxFloat(sender)`. 
```
"notificiation" -> "notification"

yondonfu commented on doc/database.md:
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
+
+## Table `ticketQueue`
+
+**orchestrator/redeemer** only. `ticketQueue` Tracks winning tickets for probabilistic micropayments.
```
Capitalize orchestrator and redeemer to match style used in the rest of this document (for example, note the "Orchestrator only" phrase for the `winningTickets` table section).

yondonfu replied:
Bump

kyriediculous replied:
fixed in 17af7bb

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+
+
+## Networking 
```
Let's add this section to [doc/networking.md](https://github.com/livepeer/go-livepeer/blob/master/doc/networking.md) following the format used there so that all networking docs are in one place. We can link to the relevant section of the networking docs from this document so that someone reading this document to understand the ticket redemption service flow can easily find networking information for the ticket redemption service.

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+\* In a multi-O setup `MonitorMaxFloatUpdate` is sent to all orchestrators in the cluster and a local cache is kept, even if the orchestrator is not currently working with the broadcaster but another orchestrator in the cluster is. This allows a broadcaster to smoothly fail over to other orchestrators within the same multi-O cluster but also creates additional storage and networking overhead. Unu ...
+
+![Ticket Flow](https://i.imgur.com/jd83N3c.png)
```
IIUC, this diagram indicates that the Redeemer only calls SubscribeMaxFloatChange for a B address when the Redeemer queues a winning ticket from that B and the Redeemer is not yet aware of that B. This would mean that an O would not receive max float updates for a B that is sending it payments until O receives a winning ticket from B. In this scenario, if B's max float changes due to a ReserveFunded or NewRound event, O would not receive those updates which impacts the ticket params/pricing that it advertises to B.

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+# Ticket Redemption Service
+
+The ticket redemption service allows an Orchestrator to run a separate service for redeeming winning tickets on-chain using a separate ETH account. 
```
> allows an Orchestrator to run a separate service

It might get confusing with "orchestrator-the-node" versus "orchestrator-the-operator", since redemption services are not going to be orchestrators.

> using a separate ETH account.

Maybe clarify "separate ETH account on the redemption service". 

Altogether, this could be tweaked to something like: "The ticket redemption service allows orchestrator nodes to redeem winning tickets on-chain using a separate ETH account on the ticket redemption service. Orchestrator nodes no longer need to have an account with ETH to pay for transactions."

kyriediculous replied:
This should be fixed

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+The ticket redemption service allows an Orchestrator to run a separate service for redeeming winning tickets on-chain using a separate ETH account. 
+
+The ticket redemption service is responsible for redeeming winning tickets as well as pushing max float updates about broadcasters back to the orchestrator(s). 
```
Is there a doc anywhere that describes the particulars of the PM mechanism, specifically the max float? It might be helpful to link to that here, so readers understand what a max float is.

kyriediculous replied:
There is this old open PR but it doesn't mention max float https://github.com/livepeer/go-livepeer/pull/1099/files

yondonfu replied:
The [PM spec](https://github.com/livepeer/wiki/blob/master/spec/streamflow/pm.md#reserve) provides a description of the float concept. We could add link in the redeemer doc for the reader's reference.

kyriediculous replied:
I added a note that links to that section of the pm spec

j0sh reviewed (changes requested):
It would be good to also have usage instructions, maybe after the introduction, with recommended CLI flags for both the redeemer service and the orchestrator node.

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+1. The `ticketQueue` is a loop that runs everytime a new block is seen. It will then pop tickets off the queue starting with the oldest ticket first, and sends it to the `LocalSenderMonitor` for redemption if the `recipientRand` for the ticket has expired. 
+
+2. When the `LocalSenderMonitor` receives a ticket from the `ticketQueue` it will substract `ticket.faceValue` from the outstanding `maxFloat` as long as the ticket is in limbo. 
```
Clarify "in limbo" ? Would it be clearer to say "not yet confirmed on the blockchain" ?

kyriediculous replied:
pending between the Redeemer and the ethereum node 

This can be either pending on-chain or an HTTP request to the eth node that has some latency to it.

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+\* In a multi-O setup `MonitorMaxFloatUpdate` is sent to all orchestrators in the cluster and a local cache is kept, even if the orchestrator is not currently working with the broadcaster but another orchestrator in the cluster is. This allows a broadcaster to smoothly fail over to other orchestrators within the same multi-O cluster but also creates additional storage and networking overhead. Unu ...
+
+![Ticket Flow](https://i.imgur.com/jd83N3c.png)
```
If an image is copy-pasted into a github text box, github will upload it automatically to `githubusercontent.com` and present a URL to the image. Using that URL here is probably a bit better than cross-linking to imgur. For example, here's the same image hosted on Github: https://user-images.githubusercontent.com/292510/84299949-776ce480-ab06-11ea-9a88-fa052dc15c85.png

kyriediculous replied:
I added an assets folder instead so that both the exported PNG as well as the source for the sequence diagram can be part of the repo.

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+1. On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server for `maxFloat` updates from broadcasters the orchestrator (or all orchestrators in a multi-orchestrator setup) is working with.* 
+
+2. Anytime a winning ticket is submitted to the `Redeemer Server`, the `Redeemer Server` will monitor it's own `LocalSenderMonitor` for `maxFloat` change notifications for the sender of the ticket using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` (if not already watching for updates from that particular sender). 
```
Can we clarify what a "sender" is here for folks unfamiliar with the code, or use more direct terminology, like "broadcaster" ?

kyriediculous replied:
yeah I guess the question is whether we want to use the language of the `pm` protocol or the livepeer protocol.

Perhaps we can add a note that currently sender is interchangeable with broadcaster

kyriediculous replied:
fixed in 8a6184d1

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+## Monitoring Max Float
+
+1. On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server for `maxFloat` updates from broadcasters the orchestrator (or all orchestrators in a multi-orchestrator setup) is working with.* 
```
Would it be clearer to say, 

"On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server. Each time the redeemer receives a winning ticket, all subscribed orchestrators receive a `maxFloat` update for the broadcaster that generated the winning ticket. All orchestrators receive this update, regardless of whether the orchestrator is working with the broadcaster."

"Within the `Redeemer Server`, float monitoring works in the following way: ..." 

Then points #2-4 can describe the redeemer's local float monitoring in more detail. (Should we also mention how orchestrator-local float monitoring works in regular mode?)

kyriediculous replied:
this has been removed

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,81 @@
+
+During discovery (`GetOrchestratorInfo`), an orchestrator invokes the `MaxFloat` RPC call. Currently it **does not cache** the returned value but uses it to calculate the `TicketParams` for the `GetOrchestratorInfo` response, then discards it. 
+
+This is done to not start a watchdog and keep local cache for each broadcaster that discovers our node. 
```
"This is done to not start a watchdog and keep local cache..." is a bit hard to parse.  I wonder if the mention of "no caching" above is sufficient, and this line can be omitted? Or maybe go a bit further into the reasons we don't keep a cache. Eg, "A cache is not kept because the `GetOrchestratorInfo` request is only informational, and does not guarantee work will actually come in. Hence, we avoid persisting local state after each request, such as a cache."

kyriediculous replied:
this has been removed

kyriediculous commented:
this branch has been force pushed with changes in accordance with https://github.com/livepeer/go-livepeer/pull/1513#issuecomment-644416131

Commit:
# This is a combination of 2 commits.
# This is the 1st commit message:

doc: redeemer sequence diagrams source

# This is the commit message #2:

doc: redeemer sequence diagrams

yondonfu commented on doc/networking.md:
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
+
+### gRPC `MonitorMaxFloat`: `MaxFloatReq` -> stream `MaxFloatUpdate`
+
+`MonitorMaxFloat` is a server-side streming RPC method that is called by an Orchestrator to receive indirect max float updates for a sender. The stream will remain open until either the client or the server closes the stream or connection. 
```
I think we can drop "indirect".

kyriediculous replied:
fixed in c29f76c

yondonfu commented on doc/networking.md:
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
+
+`MonitorMaxFloat` is a server-side streming RPC method that is called by an Orchestrator to receive indirect max float updates for a sender. The stream will remain open until either the client or the server closes the stream or connection. 
+
+The request follows the same format as the `Maxfloat` method:
```
`Maxfloat` -> `MaxFloat`.

kyriediculous replied:
fixed in  c29f76c

yondonfu commented on doc/networking.md:
````diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
+}
+```
+
+The response message also follows the same format but is now _stream_  of `MaxFloatUpdate` messages: 
````
"_stream_" -> `a _stream_`.

kyriediculous replied:
fixed in c29f76c

yondonfu commented on doc/networking.md:
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
 unless the intent is to break backwards compatibility.
+
+
+# Redeemer networking spec 
```
Let's move this section after the "Ping" section so that it is grouped together with the rest of the networking flow sections. A more descriptive name for this section may be "Remote Ticket Redemption".

kyriediculous replied:
But it's not necessarily part of the core livepeer networking protocol, should we keep those separate?

yondonfu replied:
There are distinct sections for "Broadcaster to Orchestrator" and "Broadcaster to Transcoder" (and there should probably be a "Orchestrator to Transcoder" section added at some point) so adding an additional, but separate section seems reasonable to me. Given the section naming convention, maybe a better section name would be "Orchestrator to Redeemer".

kyriediculous replied:
fixed in c29f76c

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,56 @@
+
+It is responsible for redeeming winning tickets as well as pushing _max float_ updates about broadcasters back to its connected Orchestrators.
+
+_Max float_ is the guaranteed value an Orchestrator will be able from a Broadcaster's reserve. It accounts for the current reserve allocation from a Broadcaster to an Orchestrator as well as pending winning ticket redemptions. 
```
"able from a Broadcaster's reserve" -> "able to claim from a Broadcaster's reserve"

kyriediculous replied:
fixed in 9f594b4

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,56 @@
+
+_If this call fails the response from step 1 is returned, but not kept in cache to prevent duplicate streams_
+
+3. The `RedeemerServer` go-routine started by the RPC call in step 2 will start a subscription to listen for max float changes from the `LocalSenderMonitor` for the specified `sender` using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)`.
```
"go-routine" -> "goroutine"

kyriediculous replied:
fixed in 1110b4d91bfe9ad1edfb4f658bcd6e2c89e03090

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,56 @@
+
+## Monitoring Max Float
+
+1. When max float for a `sender` is requested from the `RedeemerClient` but no local cache is available, an (unary) RPC call will be sent to the `RedeemerServer`. 
```
The actual implementation is called `Redeemer` so let's replace `RedeemerServer` with `Redeemer`?

kyriediculous replied:
fixed in fixed in 9f594b4

yondonfu commented on doc/networking.md:
````diff
@@ -162,6 +162,65 @@ message PingPong {
+
+```
+message MaxFloatUpdate {
+    bytes max_float= 2;
````
Let's update this to match the definition that was updated as a part of this [comment](https://github.com/livepeer/go-livepeer/pull/1513#discussion_r441216728).

kyriediculous replied:
fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

yondonfu commented on doc/networking.md:
````diff
@@ -162,6 +162,65 @@ message PingPong {
+
+```
+message MaxFloatUpdate {
+    bytes max_float= 2;
````
Let's update this to match the definition that was updated as a part of this [comment](https://github.com/livepeer/go-livepeer/pull/1513#discussion_r441216728).

kyriediculous replied:
fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

yondonfu commented on doc/networking.md:
````diff
@@ -162,6 +162,65 @@ message PingPong {
+}
+```
+
 ### Notes
````
I think these notes are supposed to be a part of the "Ping" section.

kyriediculous replied:
fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

yondonfu commented on doc/redeemer.md:
```diff
@@ -0,0 +1,58 @@
+
+2. A second RPC call to `MonitorMaxFloat(sender)` will open up a server-side gRPC stream to receive future update. 
+
+_If this call fails the response from step 1 is returned, but not kept in cache to prevent duplicate streams_
```
Is this referring to step 2? If so, why would keeping the result from step 1 in the cache result in duplicate streams? If step 2 fails, wouldn't there be no stream at all?

kyriediculous replied:
Ah yes that's true, this must be an old line that I repurposed. 

I think it should state something like "to prevent stale cache due to not being able to receive further updates"

kyriediculous replied:
fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

yondonfu reviewed (commented):
Changes look good to me. 👍   for a rebase from me as long as comments from other reviewers are addressed.

j0sh commented on doc/redeemer.md:
```diff
@@ -0,0 +1,60 @@
+
+8. The local cache for `sender` will be cleaned up if is not requested for 5 minutes. 
+
+![Ticket Flow](./assets/redeemer/ticketflow.png)
```
I really like these sequence diagrams, but 200-300 KB is kind of large. I suppose if we're going to be doing a lot more of these types of checkins, we should look at a blob storage extension such as git-lfs (which Github supports).

kyriediculous replied:
Do you want to include this in this PR? I think this is the only remaining issue for this PR, right?

j0sh replied:
Up to you, but I'm generally mindful of repo growth especially with large binary blobs in Git. Another option: since the source for these diagrams has been checked in and can be reproduced [1], we could upload the images themselves to Github (drag+drop the image in this text box) and link within the docs?

[1]  Maybe add a comment to the source(s) noting which tools / service can interpret the code for these diagrams?

kyriediculous replied:
now using git lfs for `.png` files

kyriediculous committed:
gitattributes

kyriediculous committed:
doc: redeemer sequence diagrams

yondonfu reviewed (approved):
LGTM
